### PR TITLE
Reduce memory churn when logging

### PR DESCRIPTION
**Pull request by ktf**

### CODE REVIEW

**Comment by ktf** [comment]
If I understand correctly, it currently creates a bunch of heap allocated strings from things which are guaranteed to have a long lifetime and it does so for each invocation of `LOG()` or `LOGP()`. Am I missing something trivial?

**Comment by rbx** [comment]
I think this is a nice change in principle. Indeed, I don't see the reason to recreate that metadata content if the original object lives long enough.

This would however be a breaking change, as the user of LogMetaData (custom sink user) would have to adjust from `std::string` to `const char*`. As you immediately see in the failing test because of:
https://github.com/FairRootGroup/FairLogger/blob/master/test/sinks.cxx#L87
which, with this PR, is comparing different addresses and thus failing.

**Review by dennisklein — COMMENTED** [review]
:+1: for the optimization, I also think it is a good idea. While at it, let's bump the minimum C++ level from 14 to 17 here https://github.com/FairRootGroup/FairLogger/blob/56b90a7445dcdaa85bf92a89905df0bf72796cb0/cmake/FairLoggerLib.cmake#L97 and use `std::string_view`s instead of `const char*`s.

**Comment by ktf** [comment]
> and use std::string_views instead of const char*s.

This would result in `strlen` being called on the construction of the `string_view` (because `__file__` / `source_location::file_name` are a char const*).

IMHO, there is nothing bad in using a char const*: the implication is simply that the lifetime of the metadata must be longer than the one of the logger and the logger does not own it. I would be suprised this was not the case, given the metadata is really the context of the log operation and it has by definition longer lifetime than the logger itself.

**Comment by ktf** [comment]
On a second thought, we could have string_views for the known strings, like severity_map. Changing that.

**Comment by ktf** [comment]
I updated the PR according to the discussion, however I still see an error badge but I cannot find the error, could you check?

**Comment by dennisklein** [comment]
> > and use std::string_views instead of const char*s.
> 
> This would result in `strlen` being called on the construction of the `string_view` (because `__file__` / `source_location::file_name` are a char const*).
> 

I have never verified it, but I would hope this has no runtime cost for the above cases, no?

**Comment by dennisklein** [comment]
> I updated the PR according to the discussion, however I still see an error badge but I cannot find the error, could you check?

https://cdash.gsi.de/index.php?project=FairLogger, e.g. https://cdash.gsi.de/viewBuildError.php?buildid=457734

Basically, one needs to also bump to `cxx_std_17` here
https://github.com/FairRootGroup/FairLogger/blob/56b90a7445dcdaa85bf92a89905df0bf72796cb0/CMakeLists.txt#L57

My bad, sry.

**Comment by dennisklein** [comment]
> > > and use std::string_views instead of const char*s.
> > 
> > 
> > This would result in `strlen` being called on the construction of the `string_view` (because `__file__` / `source_location::file_name` are a char const*).
> 
> I have never verified it, but I would hope this has no runtime cost for the above cases, no?

https://godbolt.org/z/TeY9rG5aj - Only the 4th case (`log_()`) calls `strlen`, no?

and I got another insight, `__LINE__` and `source_location::line()` are integers, we probably should store it as an int then, no?

**Comment by ktf** [comment]
> I have never verified it, but I would hope this has no runtime cost for the above cases, no?

yes, you are of course right, the issue was that I was still passing it as char const*, so inside the Logger constructor it had to assume it was a runtime string. For __FILE__ and co it should indeed be ok. This will still hold in general though. Do you mean that using it with anything but __FILE__ & co is probably not expected? Also in that case, I would still prefer the `char const*` because it needs 8 bytes less per metadata item.

I could change `file` though, given one needs to do the reverse find in any case. I will do that.

**Comment by dennisklein** [comment]
> Do you mean that using it with anything but FILE & co is probably not expected? Also in that case, I would still prefer the `char const*` because it needs 8 bytes less per metadata item.

yes, this is definitely the expected primary use case. Are those 8 Bytes really relevant? Or can we favor safer-to-use string APIs instead?

**Comment by ktf** [comment]
Well, with a bit of extra effort one could have LogMetaData below a cacheline, which probably does not harm. Let me reverse your question: what is the usecase for the safer-to-use string_view API? In which case would it come handy?

**Comment by ktf** [comment]
and if it does come handy, why not simply moving the char const* as private members and expose proper getters?

**Comment by dennisklein** [comment]
> Well, with a bit of extra effort one could have LogMetaData below a cacheline, which probably does not harm. Let me reverse your question: what is the usecase for the safer-to-use string_view API? In which case would it come handy?

Maintainability, unfortunately, it is easy to get C++ code wrong, e.g.

```
        fInfos.process_name = fProcessName.c_str();
        fInfos.file = file.data() + pos + 1;
```

Stuff like this (pointer arithmetic) is just the risk*ier* choice. One has to read extra carefully multiple lines to not break this in the future.

And I know the argument, it is for ultimate (in this case more speculatively better :P) performance and its inside a library and libraries are allowed to take more risks (because library authors are better programmers...). The final decision lies with @rbx anyways. I personally would have used string_views for the strings and an appropriate int type for the line number.

**Comment by rbx** [comment]
I much more prefer ints for the lines too, can be a shorter one if it fits. For the strings I don't have a strong preference. Can we show in benchmark that it matters?

File: logger/Logger.cxx [hunk 1]
```diff
@@ -144,16 +145,15 @@ map<Verbosity, VSpec> Logger::fVerbosities =
...
 {
     if (!fIsDestructed) {
         size_t pos = file.rfind("/");
-
         // fInfos.timestamp is filled conditionally
         // fInfos.us is filled conditionally
-        fInfos.process_name = fProcessName;
-        fInfos.file = file.substr(pos + 1);
+        fInfos.process_name = fProcessName.c_str();
+        fInfos.file = file.substr(pos + 1).data();
```
Comment (dennisklein) [inline logger/Logger.cxx]: I guess converting `string_view::data()` to a null-terminated "C-String" is an error in any case. (If you look at the diff, one can see the relevant hint, `data() != c_str()` in terms of null-termination guarantees) Better to keep the `char const * file` argument then as it - if the reader knows about the C-String-World - documents clearly that we expect a null-terminated string. By accepting a `string_view` the interface does not express this assumption/requirement any more.

**Comment by ktf** [comment]
> fInfos.file = file.data() + pos + 1;

This is indeed unnecessary. I changed it to be what was before (almost). That said, I frankly do not buy the "better maintainability" of strings. If we are here discussing for half a day it's because I have an hot spot in the profiler which comes from non printed messages which allocated "maintainable" strings. Maybe we save some (hypothetical) maintainance cost coming from pointer arithmetics, but then we pay the "maintainance" cost of undetected implicit constructors. Personally, I much prefer the former. 

To me the rationale should be:

* If you need the size, use a str::string_view.
* If you need it to be dynamic, use a std::string.
* If you simply need to stream it until you reach `\0`, use a const char*.

That said, in the interest of getting rid of the allocations, I will change everything to use a std::string_view and then we discuss about the impact and possible performance optimisation (can i import catch2 for some benchmarking?).

@rbx For the line, i thought about that, my worry is that formatting the number would be too expensive. I also fear that the current CONVERTTOSTRING creates a bunch of small static strings. Although maybe the compiler is smart enough to put them on the stack....

**Comment by dennisklein** [comment]
> can i import catch2 for some benchmarking?

Yes, from my side that would be fine. And since it is a private build(-and-test)-time only dependency, I also don't mind integration via FetchContent as described here: https://github.com/catchorg/Catch2/blob/devel/docs/cmake-integration.md#cmake-targets

File: logger/Logger.cxx (hunk 1 above)
Comment (ktf) [reply logger/Logger.cxx]: Fair point. For now I simply use a string view everywhere.

**Comment by ktf** [comment]
Can you please advise on the remaining errors? I cannot find the actual message anywhere.

**Comment by dennisklein** [comment]
> Can you please advise on the remaining errors? I cannot find the actual message anywhere.

See https://github.com/FairRootGroup/FairLogger/pull/57#issuecomment-2126553175

**Review by rbx — APPROVED** [review]

**Comment by ktf** [comment]
Thanks. Can I have a release with it?

**Comment by rbx** [comment]
> Thanks. Can I have a release with it?

Done. https://github.com/FairRootGroup/FairLogger/releases/tag/v2.0.0

**Pull request merged**